Improvements

Aurelien30000 opened the pull request:
---

### Get rid of hardcoded world heights

- Fix inability to clamp Y outside of Y 0 -> 256.
- Get rid of hardcoded world heights in brushes.

---

### Fixes and refactoring (deprecated methods, inconsistent format, etc)

- Utilize ``LivingEntity#addPotionEffects`` in PunishBrush, multiple effects of the same type are now supported.
- Propagate a common and _pleasant_ format to messages containing lists.
- Don't tab-complete commands if sender doesn't have the required permission.
- Don't tab-complete brushes the player doesn't have access to.
- Replace useless Bukkit Vectors usage.
- Use ``com.sk89q.worldedit.util.Direction`` instead of direct vectors.
- Fix stencil brushes sometimes failing even with correct arguments.
- Fix scanner brush not scanning towards to right direction.
- Fix error when trying to tab-complete from console.
- Remove hardcoded limits and default values in command messages.
- Create default storage folders on startup, such as stencils and signs.

---

### Major improvements to commands and configuration:

- Add a command to reload config requiring an admin permission ``voxelsniper.admin``.
- Make some general fields configurable, such as default flag size or warning threshold.
- Improve brush scalability adding configurable values AND default-values to some brushes.
- Implement config migration for future changes.
- Informational commands such as /vs brushes, info, perf, perflong or reload are now executable from console and players.
- Add ``BrushProperties`` field to all brush instances, their name and properties should be accessible.

---

Aurelien30000 committed:
Get rid od hardcoded world heights
- Fix inability to use brushes outside of Y 0 -> 256

Aurelien30000 committed:
Get rid of hardcoded world heights in brushes

Aurelien30000 committed:
Remove use of deprecated method:
- Utilize ``LivingEntity#addPotionEffects`` in PunishBrush, multiple effects of the same type are now supported.

Aurelien30000 committed:
Improve some commands:
- Propagate a common and _pleasant_ format to messages containing lists.
- Don't tab-complete commands if sender doesn't have the required permission.
- Don't tab-complete brushes the player doesn't have access to.

Aurelien30000 committed:
A few improvements:
- Replace useless Bukkit Vectors usage.
- Use ``com.sk89q.worldedit.util.Direction`` instead of direct vectors.

Aurelien30000 committed:
Improve config:
- Add a command to reload config requiring an admin permission ``voxelsniper.admin``.
- Make some general fields configurable such as default flag size or warning threshold.

Aurelien30000 committed:
Improve brush scalability:
- Add configurable values AND default-values to some brushes.
- Fix some issues parsing BlockType for example.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/stencil/StencilBrush.java:
```diff
@@ -35,10 +34,11 @@
  */
 public class StencilBrush extends AbstractBrush {
```
Using a pipe as delimiter for words ending with an `l` are quite hard to read imo

Also, how does one use this tool? I currently have no stencil in place, so I used `/b sl ffs`, Favs tells me I'm missing arguments. According to info parameter, this is the correct syntax. If the file doesn't exist, Favs should let the actor know about it, if that's the case here.

Aurelien30000 replied:
> Using a pipe as delimiter for words ending with an l are quite hard to read imo
Oof, whole format should be changed everywhere if changed here no?

Ty, you found a bug here.

NotMyFault replied:
> Oof, whole format should be changed everywhere if changed here no?

For pipes?

Aurelien30000 replied:
yes

NotMyFault replied:
I only spotted this one yet. Other cases may be fine?

Aurelien30000 replied:
Right, other cases may be fine, ``l|`` is only used in Stencil brushes.

NotMyFault replied:
👍🏻

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/command/executor/VoxelSniperExecutor.java:
```diff
@@ -106,10 +112,43 @@ public void executeCommand(CommandSender sender, String[] arguments) {
                 sender.sendMessage("FastAsyncVoxelSniper is " + (sniper.isEnabled() ? "enabled" : "disabled"));
                 return;
+            } else if (firstArgument.equalsIgnoreCase("info")) {
+                if (sender.hasPermission("voxelsniper.admin")) {
```
That command doesn't need a permission.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/command/executor/VoxelSniperExecutor.java:
```diff
@@ -106,10 +112,43 @@ public void executeCommand(CommandSender sender, String[] arguments) {
+            } else if (firstArgument.equalsIgnoreCase("reload")) {
+                if (sender.hasPermission("voxelsniper.admin")) {
+                    plugin.reload();
+                    sender.sendMessage(ChatColor.GREEN + "FastAsyncVoxelSniper reloaded!");
```
Maybe leave a note that the config has been reloaded, not the plugin.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/config/VoxelSniperConfig.java:
```diff
@@ -1,29 +1,66 @@
+    private final int defaultCylinderCenter;
+    private final Map<String, Map<String, Object>> brushProperties;
 
     public VoxelSniperConfig(
```
Javadocs

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/config/VoxelSniperConfig.java:
```diff
@@ -1,29 +1,66 @@
+        this.brushProperties = brushProperties;
     }
 
     public boolean isMessageOnLoginEnabled() {
         return this.messageOnLoginEnabled;
     }
 
+    public BlockType getDefaultBlockMaterial() {
+        return defaultBlockMaterial;
+    }
+
+    public BlockType getDefaultReplaceBlockMaterial() {
+        return defaultReplaceBlockMaterial;
+    }
+
+    public int getDefaultBrushSize() {
+        return defaultBrushSize;
+    }
+
     public int getLitesniperMaxBrushSize() {
         return this.litesniperMaxBrushSize;
     }
```
Javadocs

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/config/VoxelSniperConfig.java:
```diff
@@ -32,4 +69,25 @@ public int getLitesniperMaxBrushSize() {
         return this.litesniperRestrictedMaterials;
     }
 
+    public int getBrushSizeWarningThreshold() {
+        return brushSizeWarningThreshold;
+    }
+
+    public int getDefaultVoxelHeight() {
+        return defaultVoxelHeight;
+    }
+
+    public int getDefaultCylinderCenter() {
+        return defaultCylinderCenter;
+    }
+
+    public Map<String, Map<String, Object>> getBrushProperties() {
+        return brushProperties;
+    }
+
+    public void saveBrushPropertyToConfig(String brush, String propertyKey, Object value) {
+        plugin.getConfig().set(VoxelSniperConfigLoader.BRUSH_PROPERTIES + "." + brush + "." + propertyKey, value);
+        plugin.saveConfig();
+    }
+
```
Javadocs

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/config/VoxelSniperConfigLoader.java:
```diff
@@ -25,7 +47,7 @@ public VoxelSniperConfigLoader(FileConfiguration config) {
-     * Returns if the login message is enabled.
+     * Return if the login message is enabled.
      *
      * @return true if message on login is enabled, false otherwise.
```
If javadocs contain (boolean) operational expressions, please highlight them using a code tag, e.g 
`@return {@code true} if message on login is enabled, {@code false} otherwise.`.
The difference is small, but noticeable if you browse Javadocs on the interwebz or in your IDE
Current:
![](https://i.imgur.com/T4xiREe.png)
![](https://i.imgur.com/L63DDHV.png)

With tags:
![](https://i.imgur.com/l3gvz3b.png)
![](https://i.imgur.com/XtDEyPE.png)

Might be just personal preference, but the latter is easier to read.

Aurelien30000 replied:
Ty very much for info and details ;)

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/config/VoxelSniperConfigLoader.java:
```diff
@@ -43,7 +65,73 @@ public void setMessageOnLoginEnabled(boolean enabled) {
-     * Returns maximum size of brushes that LiteSnipers can use.
+     * Return default block material.
+     *
+     * @return default type
```
*Return default block material type? I guess

This affects the following docs too.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/performer/Performer.java:
```diff
@@ -2,14 +2,53 @@
+    /**
+     * Initialize performer.
+     *
+     * @param snipe Snipe
```
Eh?

Aurelien30000 replied:
Forgot a word, but actually this method initialize data needed by ``perform`` method (local fields, empty lists, etc).

NotMyFault commented on src/main/resources/config.yml:
```diff
@@ -1,3 +1,186 @@
 message-on-login-enabled: true
```
Maybe put an immutable revision at the top, so incase we add new entries or deprecate existing ones, we can do config upgrades based on the version the user has?

NotMyFault replied:
This should also be set somewhere in the code, in case if someone removes or changes it.
Comment can be like `# The first line is immutable. Don't change the value, it'll reset itself after every reboot.`

NotMyFault commented on src/main/resources/config.yml:
```diff
@@ -1,3 +1,186 @@
+default-brush-size: 3
 litesniper-max-brush-size: 5
-litesniper-restricted-materials:
+litesniper-restricted-materials: []
```
Add a default couple like barrier and bedrock or so to show the format being used.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/config/VoxelSniperConfigLoader.java:
```diff
@@ -1,19 +1,41 @@
 public class VoxelSniperConfigLoader {
 
     private static final String MESSAGE_ON_LOGIN_ENABLED = "message-on-login-enabled";
+    private static final String DEFAULT_BLOCK_MATERIAL = "default-block-material";
+    private static final String DEFAULT_REPLACE_BLOCK_MATERIAL = "default-replace-block-material";
+    private static final String DEFAULT_BRUSH_SIZE = "default-brush-size";
     private static final String LITESNIPER_MAX_BRUSH_SIZE = "litesniper-max-brush-size";
     private static final String LITESNIPER_RESTRICTED_MATERIALS = "litesniper-restricted-materials";
+    private static final String BRUSH_SIZE_WARNING_THRESHOLD = "brush-size-warning-threshold";
+    private static final String DEFAULT_VOXEL_HEIGHT = "default-voxel-height";
+    private static final String DEFAULT_CYLINDER_CENTER = "default-cylinder-center";
```
Where are the added entries in the config?

Aurelien30000 replied:
```yaml
message-on-login-enabled: true
default-block-material: AIR
default-replace-block-material: AIR
default-brush-size: 3
litesniper-max-brush-size: 5
litesniper-restricted-materials: [BARRIER, BEDROCK]
brush-size-warning-threshold: 20
default-voxel-height: 1
default-cylinder-center: 0
```

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/stencil/StencilBrush.java:
```diff
@@ -83,7 +90,7 @@ public void handleCommand(String[] parameters, Snipe snipe) {
             try {
                 this.filename = parameters[1 + pasteParam];
-                File file = new File(VoxelSniperPlugin.getPlugin().getDataFolder() + "/stencils/" + this.filename + ".vstencil");
+                File file = new File(plugin.getDataFolder() + "/stencils/" + this.filename + ".vstencil");
```
Maybe do it WE-like and create a folder on initial setup if it doesn't exist, so the user knows where to put the files.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/util/message/Messenger.java:
```diff
@@ -45,7 +45,7 @@ public void sendReplaceBlockDataMessage(BlockState replaceBlockData) {
 
     public void sendBrushSizeMessage(int brushSize) {
         sendMessage(ChatColor.GREEN + "Brush Size: " + ChatColor.DARK_RED + brushSize);
```
Is the input validated somewhere? Currently you can use brushes with a size lesser than 0, where 0 is already the smallest size you can have.

Aurelien30000 replied:
Should be validated in ``BrushExecutor.java`` and in config loader, this line just takes a valid input to print it.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/AbstractBrush.java:
```diff
@@ -61,7 +72,7 @@ public void handleCommand(String[] parameters, Snipe snipe) {
      * @return Sorted completions.
      */
-    public List<String> sortCompletions(Stream<String> completions, String parameter, int index) {
+    protected List<String> sortCompletions(Stream<String> completions, String parameter, int index) {
```
Is there a reason why all methods have been changed to protected in this class? Wouldn't that result in everyone using the API no longer being able to use these methods (unless they aren't considered API).

Aurelien30000 replied:
What did I do 🤯 ty

Aurelien30000 commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/FlatOceanBrush.java:
```diff
@@ -16,8 +16,14 @@
+
+    @Override
+    public void loadProperties() {
+        this.waterLevel = getIntegerProperty("default-water-level", DEFAULT_WATER_LEVEL);
```
Values are not updated because brushes are stored in the toolkit, unbind it and assign it again weird stuff here, reload is just supposed to take account of new values _later_ imo.
No idea about the +1, that was always been the case.

NotMyFault replied:
We can likely address the +1, from what I've seen it serves no purpose.

Aurelien30000 replied:
Humm, just noticed and wondered, there is no "height" in FlatOceanBrush, wdym exactly, +1 only adds a layer of water on this brush. Are you somehow talking about another one?

NotMyFault replied:
I'll get back to that later.
![](https://i.imgur.com/WMiotFa.jpg)
Is this how the brush is supposed to look like? yo 5 & yl 3
Looks alike the canyon brush just in weird? I imagined it would act like any other brush and putting down the blocks on top of the block you target on, instead of erasing a whole chunk.

Aurelien30000 replied:
Oceans brushes currently recreate the whole terrain indeed, maybe to add to our to-do list?

NotMyFault replied:
Likely, removing the entire chunk is eh.. weird to place a few rows of water and dirt. Can also add to the list to create an option to specify material instead of classic dirt.

Aurelien30000 committed:
Fix other hardcoded world limits

Aurelien30000 commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/RulerBrush.java:
```diff
@@ -5,21 +5,24 @@
 import java.util.List;
 import java.util.stream.Stream;
 
 public class RulerBrush extends AbstractBrush {
```
Select a block with arrow, then use gunpowder on another block to measure distance between both.

OR 

Use ``/b ruler <xoff> <yoff> <zoff>`` to define offset. Once set to something else than 0, use the arrow on a block, it places a block with entered offsets.

Aurelien30000 commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/ScannerBrush.java:
```diff
@@ -17,11 +16,23 @@
 public class ScannerBrush extends AbstractBrush {
```
Checks completely messed, broken brush, will fix it, ty.

Aurelien30000 commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/performer/PunishBrush.java:
```diff
@@ -29,24 +29,42 @@
 
 public class PunishBrush extends AbstractPerformerBrush {
```
You can't apply several punishments https://voxelsniper.fandom.com/wiki/VoxelSniper_Wiki#Punish_Brush
Let's add this feature to our ideas list :)

Aurelien30000 commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/performer/SetBrush.java:
```diff
@@ -15,6 +15,13 @@
     private BlockVector3 block;
     private World world;
 
+    private int selectionSizeMax;
```
They do the same job, gunpowder uses ``lastBlock`` instead of ``targetBlock``.

Aurelien30000 commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/shell/ShellSetBrush.java:
```diff
@@ -17,6 +17,13 @@
 
     private static final int MAX_SIZE = 5000000;
```
🤔 Don't know exactly, and found no doc about this...

![2021-08-28_17 47 51](https://user-images.githubusercontent.com/43724816/131223414-498159d0-bf83-404a-8d72-d874d6857fc2.png)

Aurelien30000 commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/performer/ThreePointCircleBrush.java:
```diff
@@ -18,6 +18,8 @@
 
 public class ThreePointCircleBrush extends AbstractPerformerBrush {
```
Another uppercase problem ;)
_A list, don't know, added a list to all use of enums or registries 😅 _

Aurelien30000 committed:
Address requests, fix some issues

Aurelien30000 commented:
Should be good overall, just waiting for you reply on a few comments, take it easy.

Aurelien30000 committed:
Revert incorrect changes, final fixes
- Implement config migration

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/config/VoxelSniperConfigLoader.java:
```diff
@@ -28,22 +34,100 @@
+    public void updateConfig() {
+        int currentConfigVersion = getConfigVersion();
+        if (currentConfigVersion != CONFIG_VERSION_VALUE) {
+            Bukkit.getLogger().warning("Invalid config file found! Trying to apply required changes...");
```
`Bukkit#getLogger` returns the instance of the logger the server uses, not the one of the plugin.

NotMyFault commented on src/main/java/com/thevoxelbox/voxelsniper/config/VoxelSniperConfigLoader.java:
```diff
@@ -28,22 +34,100 @@
+    /**
+     * Update config settings the current config version and apply changes.
+     */
+    public void updateConfig() {
```
This can be private.

NotMyFault reviewed (changes requested):
A few things that should be addressed still:
- Fix the typos in `AbstractBrush#getProperty`
- `/b stencil <operation> <name>` throws an AIOOBE https://paste.gg/p/anonymous/2e37640f94ae4cd89a87d8aba4abe81f
Additionally to that one, leave a note whether a stencil is invalid aka when a player attempts to load a file that doesn't exist.
- Informational commands such as `/vs brushes, info, perf, perflong or reload` shouldn't be bound to players only. Using them from console should be possible too
- The methods to setup & upgrade config values can stay protected, no need to expose them.

Aurelien30000 commented on src/main/java/com/thevoxelbox/voxelsniper/brush/type/OceanBrush.java:
```diff
@@ -74,7 +74,7 @@ public void handleCommand(String[] parameters, Snipe snipe) {
                         messenger.sendMessage(ChatColor.BLUE + "Water level set to: " + ChatColor.GREEN + this.waterLevel);
                     } else {
-                        messenger.sendMessage(ChatColor.RED + "Invalid number, must be an integer greater " + this.waterLevelMin + ".");
+                        messenger.sendMessage(ChatColor.RED + "Invalid number, must be an integer >" + this.waterLevelMin + ".");
```
Ok ;)

NotMyFault commented:
If you want to address the console support for the arguments I named before, may also consider this one https://paste.gg/p/anonymous/de861cb9a5604edba3139a98d7b90916
That happens if you tab complete tools at console. Generally unneeded, they wouldn't work in the first place.
If not, convert my comment in an issue.

Aurelien30000 commented:
> If you want to address the console support for the arguments I named before, may also consider this one https://paste.gg/p/anonymous/de861cb9a5604edba3139a98d7b90916
> That happens if you tab complete tools at console. Generally unneeded, they wouldn't work in the first place.
> If not, convert my comment in an issue.
> 
> > **[pastefile1 · paste.gg](https://paste.gg/p/anonymous/de861cb9a5604edba3139a98d7b90916)**

Oh ty for pointing this out, just gonna _skip_ completions for a ``CommandSender`` that doesn't match command class requirement.

Aurelien30000 committed:
Fixes,

- Fix the typos in AbstractBrush#getProperty
- Fix /b stencil <operation> <name> throwing an AIOOBE
- Informational commands such as /vs brushes, info, perf, perflong or reload are now executable from console and players.
- Fix error when trying to tab-complete from console.

NotMyFault reviewed (approved):
LGTM, may need to adjust the pull request description to the additional changes.

Aurelien30000 commented:
Done I guess.

NotMyFault commented:
> Done I guess.

Thanks, I have drafted a release.
